### PR TITLE
Support running extension tests in browser workbench

### PR DESCRIPTION
**Pull request by aeschli**
Brings the concept of extension under deb and extension tests to the web extension host.

Additions to the workbench embedder API:
IStaticExtension:
 - new property `isUnderDevelopment`
IWorkbenchConstructionOptions:
- extensionTestsPath

The extensionDevelopmentPath is computed based on IStaticExtension.isUnderDevelopment

Alternative API would be:

IWorkbenchConstructionOptions
- staticExtensions
- staticDevExtensions
- extensionTestsPath

### CODE REVIEW

**Review by bpasero — APPROVED** [review]
I actually feel I cannot give good feedback because I am far away from this topic mentally so I trust these changes work for you.

Maybe one thing is how we want to call the API: a top level `extensionTestsPath` in the workbench construction options is maybe too prominent? Should it move as a property somewhere closer to where the extensions are provided?

**Comment by aeschli** [comment]
> I actually feel I cannot give good feedback because I am far away from this topic mentally so I trust these changes work for you.

Make sense. I wanted your opinion mostly on the embedded API, on the rest I'm quite comfortable as me and @alexdima refactored extension test running last milestone.

> Maybe one thing is how we want to call the API: a top level extensionTestsPath in the workbench construction options is maybe too prominent? Should it move as a property somewhere closer to where the extensions are provided?

Yes, I agree, it's a bit too prominent.
What about grouping it together with `driver` under a `developmentOptions` property:
```
"developmentOptions": {
    "extensionTestsPath": URI
    "driver": boolean
 }
 ```
 I'm not familiar what `driver` is for. A better name that would be also be good.

There's also the option to move the devExtensions there, which would be my preferred option:
```
"developmentOptions": {
    "extensionTestsPath": URI
    "extensions": ReadonlyArray<IStaticExtension>
    "driver": boolean
 }
 ```

**Comment by bpasero** [comment]
@aeschli I really like the `developmentOptions` thing! The `driver` option is used for the smoke tests:

https://github.com/microsoft/vscode/blob/bdf2ef774e0583ee0759179d75e5b044190a88b2/src/vs/workbench/browser/window.ts#L118-L120

I think we can easily rename that to `enableSmokeTestDriver`

File: src/vs/workbench/workbench.web.api.ts [hunk 1]
```diff
@@ -26,7 +26,7 @@ interface IResourceUriProvider {
 
 interface IStaticExtension {
 	packageJSON: IExtensionManifest;
-	extensionLocation: URI;
+	extensionLocation: UriComponents;
```
Comment (bpasero) [inline src/vs/workbench/workbench.web.api.ts:29]: Why can this not be `URI`?

File: src/vs/workbench/workbench.web.api.ts [hunk 2]
```diff
@@ -410,12 +410,31 @@ interface IWorkbenchConstructionOptions {
...
+
+	//#region development options
+
+	readonly developmentOptions?: IDevelopmentOptions;
+
+	//#endregion
+
+}
+
+interface IDevelopmentOptions {
```
Comment (bpasero) [inline src/vs/workbench/workbench.web.api.ts:423]: Make sure to export this interface at the end of the file.

**Review by bpasero — APPROVED** [review]
Yeah, looks better. Some minor feedback provided.

**Comment by aeschli** [comment]
@bpasero Sorry, just saw your comments after merging.
I saw several cases (e.g. in Joao's server) where it was missed to resolve the URLs before passing then to the web embedded API.
As there are only few places where the extensionLocation are accessed I thought it would simplify things if we loosen the API and only require UriComponents and do the resolving ourselves. 
IMO the most convenient would have been a URL sting.

yes, good catch, I will add the IDevelopmentOptions export.

**Comment by bpasero** [comment]
Yeah indeed, in retrospect any `URI` should be `UriComponents` probably...

**Comment by bpasero** [comment]
Maybe `extensionTestsPath` should also then use `UriComponents`?

**Comment by aeschli** [comment]
Oh, yes, extensionTestsPath should also be UriComponents.

**Comment by aeschli** [comment]
export IDevelopmentOptions: af76d0cf1374fd570637fe46edf3ef4d0898179a

**Comment by aeschli** [comment]
Now 2 the remaining places where URI are used in the embedded API are in the [IResourceUriProvider](https://github.com/microsoft/vscode/blob/f5b31e9a3318c30cd7aa00bc32009e55e9350a27/src/vs/workbench/workbench.web.api.ts#L23) and the similar `IExternalUriResolver`. IMO makes sense to keep URI there.

**Comment by bpasero** [comment]
Yeah no worries, I can take a look at the remaining ones. I think workspace provider also has usages and I remember Joao hit an issue there where we ignored the workspace when a `UriComponent` was passed in as workspace.

**Pull request merged**